src/Microsoft.Diagnostics.Runtime/ClrValue.cs

```diff
@@ -0,0 +1,1077 @@
+                throw new ArgumentException($"Type '{type.Name}' does not contain a field named '{fieldName}'");
+
+            if (!field.IsObjectReference)
+                throw new ArgumentException($"Field '{type.Name}.{fieldName}' is not an object reference.");
+
+            ClrHeap heap = Type.Heap;
+
+            ulong addr = ClrRuntime.IsObjectReference(ElementType) ? Object : Address;
+            addr = field.GetAddress(addr, Interior);
+            ulong obj;
+
+            if (!heap.ReadPointer(addr, out obj))
+                throw new MemoryReadException(addr);
+            
+            return new ClrObject(obj, heap.GetObjectType(obj));
+        }
+
+        /// <summary>
+        /// Retrieves a field from this value.
+        /// </summary>
+        /// <param name="name">The name of the field.</param>
+        /// <returns>A ClrValue representing this field.</returns>
+        public virtual ClrValue GetField(string name)
+        {
+            ClrElementType el = ElementType;
+            if (ClrRuntime.IsPrimitive(el))
+            {
+                // Primitives only have one field, named m_value.
+                if (name != "m_value")
+                    throw new ArgumentException(string.Format("Field '{0}' does not exist in type '{1}'.", name, Type.Name));
+
+                // Getting m_value is the same as this ClrValue...
+                return this;
+            }
+
+            if (ClrRuntime.IsObjectReference(el) || !Interior)
+                return AsObject().GetField(name);
+
+            Debug.Assert(ClrRuntime.IsValueClass(el));
+
+            ulong address = Address;
+            if (address == 0)
+                throw new NullReferenceException();
+
+            ClrType type = Type;
+            ClrInstanceField field = type.GetFieldByName(name);
+            if (field == null)
+                throw new ArgumentException(string.Format("Field '{0}' does not exist in type '{1}'.", name, Type.Name));
+
+            ulong result = field.GetAddress(address, Interior);
+            return new ClrValueImpl(_runtime, result, field);
+        }
+
+        /// <summary>
+        /// Gets a boolean field from the object.  Note that the type must match exactly, as this method
+        /// will not do type coercion.  This method will throw an ArgumentException if no field matches
+        /// the given name.  It will throw a NullReferenceException if the target object is null (that is,
+        /// if (IsNull returns true).  It will throw an InvalidOperationException if the field is not
+        /// of the correct type.  Lastly, it will throw a MemoryReadException if there was an error reading
+        /// the value of this field out of the data target.
+        /// </summary>
+        /// <param name="fieldName">The name of the field to get the value for.</param>
+        /// <returns>The value of the given field.</returns>
+        public bool GetBoolean(string fieldName)
+        {
```

Comment:
I thought of that, but didn't for two reasons:

1.  Intellisense.  Feedback was that it would be better to have named fields.
2.  C# doesn't have template specialization, so it's tough to get an efficient implementation of this.  Not impossible though, and perf it's exactly the largest concern.  So this was secondary.

I'm still a bit on the fence about this honestly, especially after seeing how similar the implementations turned out to be.